argocd: added a restrictions section (#853)
elastisys/compliantkubernetes: docs/user-guide/additional-services/argocd.md

!!! note
    Email notification will not work if the sender has 2FA enabled.

## Restrictions

Example error:

```Failed to load live state: Cluster level Namespace "application" can not be managed when in namespaced mode```

Our ArgoCD installation is using the [namespaced method](https://argo-cd.readthedocs.io/en/stable/operator-manual/installation/#non-high-availability). This means that ArgoCD has access to Roles with permissions to CRUD on objects in the inclusion list. It has a [list of namespaces](https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/#clusters) that it can look at and reconcile things every few seconds. Any feature that requires ArgoCD cluster-wide installation will not be supported with our offering.

The reason for this choice is to prevent ArgoCD from having access to objects in the inclusions list across the entire cluster. This prevents objects from being deployed onto unwanted namespaces.

> ArgoCD is not allowed to manage its own namespace. This means that features such as [Apps of Apps](https://argo-cd.readthedocs.io/en/stable/operator-manual/cluster-bootstrapping/#app-of-apps-pattern) is not available in our offering. Read more about the decision [here](../../adr/0044-argocd-managing-its-own-namespace.md).

Reviewer: Now that Apps of Apps is an alpha feature. Can we write something like
```
Apps of apps is currently an alpha feature, and customers can request it, provided they accept risks such as 

-  When Apps-of-Apps is enabled, we will not be able to provide Uptime SLAs for you on the Argo Endpoint (This will be a best effort support even in future premium environments, unless the Argo project develops the feature upstream in a more secure and isolated way)

 - Apps-of-Apps will currently be regarded as an alpha feature and, as such, we reserve the right to disable it if it interferes with our OPS policies/practices – should that unfortunate event happen, it will be in dialogue with the customer. 
 ```

Reviewer 2: Nice! Please put this into an `!!!elastisys` admonition.

    Email notification will not work if the sender has 2FA enabled.

## Restrictions

Example error:

`Failed to load live state: Cluster level Namespace "application" can not be managed when in namespaced mode`

Our ArgoCD installation is using the [namespaced method](https://argo-cd.readthedocs.io/en/stable/operator-manual/installation/#non-high-availability). This means that ArgoCD has access to Roles with permissions to CRUD on objects in the inclusion list. It has a [list of namespaces](https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/#clusters) that it can look at and reconcile things every few seconds. Any feature that requires ArgoCD cluster-wide installation will not be supported with our offering.

> The reason for this choice is to prevent ArgoCD from having access to objects in the inclusions list across the entire cluster. This prevents objects from being deployed into unwanted namespaces.

Reviewer 2: ```suggestion
The reason for this choice is that, according to the [Compliant Kubernetes mission and vision](../../mission-and-vision.md), the platform should make it hard for application developers to do the wrong thing by employing safeguards and secure defaults.
With this configuration, we prevent ArgoCD from having access to objects in the inclusions list across the entire cluster. This prevents objects from being deployed into namespaces owned by the platform administrator, which could compromise platform security and stability. For example, this choice adds another layer of protection, preventing the application developer from interfering with backups.
```

We could also link to:

- https://elastisys.io/compliantkubernetes/user-guide/demarcation/#general-principle
-

> The reason for this choice is to prevent ArgoCD from having access to objects in the inclusions list across the entire cluster. This prevents objects from being deployed onto unwanted namespaces.

Reviewer 3: nit
```suggestion
The reason for this choice is to prevent ArgoCD from having access to objects in the inclusions list across the entire cluster. This prevents objects from being deployed into unwanted namespaces.
```

The reason for this choice is that, according to the [Compliant Kubernetes mission and vision](../../mission-and-vision.md), the platform should make it hard for application developers to do the wrong thing by employing safeguards and secure defaults. With this configuration, we prevent ArgoCD from having access to objects in the inclusions list across the entire cluster. This prevents objects from being deployed into namespaces owned by the platform administrator, which could compromise platform security and stability. For example, this choice adds another layer of protection, preventing the application developer from interfering with backups. Read more about it [here](../demarcation.md#general-principle).

ArgoCD is not allowed to manage its own namespace. This means that features such as [Apps of Apps](https://argo-cd.readthedocs.io/en/stable/operator-manual/cluster-bootstrapping/#app-of-apps-pattern) does not work by default. Read more about the decision [here](../../adr/0044-argocd-managing-its-own-namespace.md).

> !!! elastisys "Apps of Apps Alpha"

Reviewer 2: ```suggestion
!!! elastisys "Apps of Apps (preview)"
```

>     Using Apps of Apps with our offering is currently an alpha feature, and customers can request it, provided they accept risks such as:

Reviewer 2: ```suggestion
    This is a preview feature. For more information, please read [ToS 9.1 Preview Features](https://elastisys.com/legal/terms-of-service/#91-preview-features).

    Using Apps of Apps with our offering is currently a preview feature, and customers can request it, provided they accept risks such as:
```

    -  When Apps-of-Apps is enabled, we will not be able to provide Uptime SLAs for you on the Argo Endpoint (This will be a best effort support even in future premium environments, unless the Argo project develops the feature upstream in a more secure and isolated way)

>     - Apps-of-Apps will currently be regarded as an alpha feature and, as such, we reserve the right to disable it if it interferes with our OPS policies/practices – should that unfortunate event happen, it will be in dialogue with the customer.

Reviewer 2: ```suggestion
    - Apps-of-Apps will currently be regarded as a preview feature and, as such, we reserve the right to disable it if it interferes with our OPS policies/practices – should that unfortunate event happen, it will be in dialogue with the customer.
```

ArgoCD is not allowed to manage its own namespace. This means that features such as [Apps of Apps](https://argo-cd.readthedocs.io/en/stable/operator-manual/cluster-bootstrapping/#app-of-apps-pattern) does not work by default. Read more about the decision [here](../../adr/0044-argocd-managing-its-own-namespace.md).

!!! elastisys "Apps of Apps (preview)"

    This is a preview feature. For more information, please read [ToS 9.1 Preview Features](https://elastisys.com/legal/terms-of-service/#91-preview-features).

    Using Apps of Apps with our offering is currently a preview feature, and customers can request it, provided they accept risks such as:

>     -  When Apps-of-Apps is enabled, we will not be able to provide Uptime SLAs for you on the Argo Endpoint (This will be a best effort support even in future premium environments, unless the Argo project develops the feature upstream in a more secure and isolated way)

Reviewer 3: nit
```suggestion
    - When Apps-of-Apps is enabled, we will not be able to provide Uptime SLAs for you on the Argo Endpoint (This will be a best effort support even in future premium environments, unless the Argo project develops the feature upstream in a more secure and isolated way)
```
